# PANDAS ASSIGNMENT

##  Submitted By: Akurathi Harshitha


## Branch: Information Technology


## Batch: Batch 2


# Output Prediction

## Q1.  Array operations and vectorization

In [ ]:
import numpy as np
import pandas as pd

arr = np.array([10, 20, 30])
s = pd.Series([10, 20, 30])

print(arr * 2)
print(s * 2)

### Predicted Output:

[20 40 60]
0    20
1    40
2    60
dtype: int64

### Similarity: 

Both NumPy arrays and Pandas Series can perform multiplication on all their elements at once without using a loop.

### Difference: 

In a NumPy array, elements are accessed using their positions. In a Pandas Series, each element has an index label, which helps us identify and access the values easily.

## Q2. Boolean Filtering

In [ ]:
import pandas as pd

df = pd.DataFrame({
    'Student': ['A', 'B', 'C', 'D'],
    'Marks': [45, 80, 65, 90]
})

result = df[df['Marks'] >= 65]

print(result[['Student', 'Marks']])

### Predicted Output:

  Student  Marks
1       B     80
2       C     65
3       D     90

### Number of Rows and Columns:

The resulting DataFrame contains 3 rows and 2 columns.

### Explanation:

The condition df['Marks'] >= 65 selects only the students who scored 65 or above. So, students B, C, and D are displayed, while student A is excluded because the marks are below 65.

## Q3.  Descriptive statistics

In [ ]:
import pandas as pd

df = pd.DataFrame({
    'Marks': [10, 20, 30, 40, 50]
})

print(df['Marks'].mean())
print(df['Marks'].max())
print(df['Marks'].median())

### Predicted Output:

30.0
50
30.0

### Explanation:

Mean: 
The mean is the average of all the marks. Here, the sum of the marks is divided by the total number of values, giving 30.0.

Maximum: 
The maximum value is the highest mark in the column, which is 50.

Median: 
The median is the middle value when all the marks are arranged in order. Here, the middle value is 30.0.

### Difference between Mean and Median:

Mean is calculated by adding all the values and dividing by the total number of values, whereas the median is the middle value of the data when arranged in order. The mean can be affected by very high or very low values, but the median is less affected by them.

## Q4.  Group-wise Analysis

In [ ]:
import pandas as pd

df = pd.DataFrame({
    'Department': ['IT', 'HR', 'IT', 'HR'],
    'Salary': [40000, 30000, 60000, 50000]
})

print(df.groupby('Department')['Salary'].mean())

### Predicted Output:

Department
HR    40000.0
IT    50000.0
Name: Salary, dtype: float64

### Explanation:

The groupby() function groups the data based on the department. It then calculates the average salary of each department using mean().

HR: (30000 + 50000) / 2 = 40000
IT: (40000 + 60000) / 2 = 50000

The output shows the average salary of each department separately.

## Q5.  Pandas Correlation - Positive

In [ ]:
import pandas as pd

df = pd.DataFrame({
    'Hours': [1, 2, 3, 4],
    'Marks': [20, 40, 60, 80]
})

print(df['Hours'].corr(df['Marks']))

### Predicted Output:

1.0

### Interpretation:

The correlation coefficient is +1.0, which indicates a perfect positive correlation between study hours and marks.

As the number of study hours increases, the marks also increase. This shows a strong positive linear relationship between the two variables.

## Q6.  Pandas Correlation - Negative

In [ ]:
import pandas as pd

df = pd.DataFrame({
    'Price': [10, 20, 30, 40],
    'Demand': [100, 80, 60, 40]
})

print(df['Price'].corr(df['Demand']))

### Predicted Output:

-1.0

### Explanation:

The correlation coefficient is -1.0, which indicates a perfect negative correlation between price and demand.

As the price increases, the demand decreases. The negative sign indicates that the two variables move in opposite directions. Since the coefficient is -1.0, the relationship is perfectly negative.

# Code Correction

## Q1. JSON and CSV    


import pandas as pd

df = pd.read_csv("students.csv", header=None)
print(df)    

A CSV file contains a header row, but the following code incorrectly reads the first student's record as column names. Correct the code

In [4]:
#Corrected code:

import pandas as pd

df = pd.read_csv("students.csv")
print(df)

    ID     Name        Date  Marks
0  101  Asha, R  2026-09-01   80.0
1  102     Ravi  01/09/2026    NaN
2  103     Neha  2026-09-03    NaN


### Explanation:

In the given code, header=None tells Pandas that the CSV file does not contain column names. Because of this, the first row, which contains the header, is treated as a data row, and Pandas assigns default column names instead.

To correct this, header=None is removed. By default, read_csv() uses the first row of the CSV file as column names, so the data is read with the correct headers.

## Q2. Data Analysis

import pandas as pd

df = pd.DataFrame({
    'Employee': ['A', 'B', 'C', 'D', 'E', 'F'],
    'Salary': [30000, 50000, 25000, 60000, 45000, 70000]
})

print(df.sort_values('Salary').head(5))                              

The objective is to display the five highest salaries from a DataFrame. Correct the code.

In [5]:
#Corrected code:

import pandas as pd

df = pd.DataFrame({
    'Employee': ['A', 'B', 'C', 'D', 'E', 'F'],
    'Salary': [30000, 50000, 25000, 60000, 45000, 70000]
})

print(df.sort_values('Salary', ascending=False).head(5))

  Employee  Salary
5        F   70000
3        D   60000
1        B   50000
4        E   45000
0        A   30000


### Explanation:

In the given code, sort_values('Salary') sorts the salaries in ascending order, from lowest to highest. Because of this, head(5) displays the first five rows, which do not contain the five highest salaries.

To correct this, ascending=False is added to sort_values(). This sorts the salaries from highest to lowest, and head(5) selects the first five rows, giving the five highest salaries.

## Q3. Cleaning Data          

import pandas as pd

df = pd.DataFrame({'Marks': [70, None, 90]})

df['Marks'].fillna(df['Marks'].mean())
print(df)    

A student wants to replace missing marks with the column's mean but writes code that does not update the DataFrame. Correct the code

In [6]:
#Corrected code:

import pandas as pd

df = pd.DataFrame({'Marks': [70, None, 90]})

df['Marks'] = df['Marks'].fillna(df['Marks'].mean())
print(df)

   Marks
0   70.0
1   80.0
2   90.0


### Explanation:

In the given code, fillna() is used to replace the missing mark with the mean of the column, but the updated result is not stored in the DataFrame. To correct this, the result of fillna() is assigned back to df['Marks']. The mean of the available marks (70 and 90) is 80, so the missing value is replaced with 80.

## Q4. Correlation

import pandas as pd

df = pd.DataFrame({
    'Hours': [1, 2, 3, 4],
    'Marks': [20, 40, 60, 80]
})

print(df.corr('Hours', 'Marks’))

The objective is to calculate the correlation between Hours and Marks. Correct the code.

In [7]:
#Corrected code:

import pandas as pd

df = pd.DataFrame({
    'Hours': [1, 2, 3, 4],
    'Marks': [20, 40, 60, 80]
})

print(df['Hours'].corr(df['Marks']))

1.0


### Explanation:

In the given code, df.corr('Hours', 'Marks') is incorrect because corr() does not accept column names as two separate arguments in this way. To calculate the correlation between Hours and Marks, we use df['Hours'].corr(df['Marks']). This calculates the correlation between the two columns and gives 1.0, indicating a perfect positive correlation.

# Edge Cases

### Q1. Two NumPy arrays have shapes (3, 1) and (1, 3). Predict the shape of their element-wise sum and explain how broadcasting works. Why might the result differ from adding two one-dimensional arrays?


In [ ]:
import numpy as np

a = np.array([[10], [20], [30]])
b = np.array([[1, 2, 3]])

result = a + b

print(result)
print(result.shape)


### 1. Predict the output of the given code.

  Result

[[11 12 13]
 [21 22 23]
 [31 32 33]]

Shape: (3, 3)

### 2. Identify the shapes of arrays a and b before addition.

Array a: Shape is (3, 1) (2D array: 3 rows, 1 column)

Array b: Shape is (1, 3) (2D array: 1 row, 3 columns)

### 3. Explain how NumPy broadcasting enables the addition of these arrays.

  NumPy compares array dimensions element-by-element starting from the trailing (rightmost) dimensions:
  1. Dimension 1 (Columns): a has size 1, and b has size 3. Since one of the dimensions is 1, a is stretched (replicated) 3 times horizontally across columns.
  
  2. Dimension 0 (Rows): a has size 3, and b has size 1. Since b has size 1, it is stretched (replicated) 3 times vertically across rows.

   Both arrays expand to an effective common shape of (3, 3) without physically copying memory:
   a behaves like: [[10 10 10
                     20 20 20
                     30 30 30]]
   b behaves like: [[1 2 3
                     1 2 3
                     1 2 3]]

   Adding them element-wise produces the final (3, 3) matrix.

### 4. What would happen if b were replaced with np.array([1, 2])? Explain the reason.
 
  If b is replaced with np.array([1, 2]), the code executes successfully without any error, producing a 2D array of shape (3, 2):
   [[11 12]
    [21 22]
    [31 32]]
    Shape: (3, 2)

Alignment: NumPy aligns array shapes starting from the rightmost (trailing) axis:
 a shape: (3, 1)
 b shape: (2,) automatically treated as (1, 2)

Broadcasting Rule: Dimensions are compatible if they are equal or if one of them is 1.
 Axis 1 (Columns): Size 1 vs 2,  a expands horizontally to size 2.
 Axis 0 (Rows): Size 3 vs 1 , b expands vertically to size 3.
 
Result: Both arrays expand to an effective shape of (3, 2), making the addition valid.

### Q2. A CSV file contains commas inside quoted text fields, missing values, and dates in different formats. Explain how incorrect parsing can affect the resulting DataFrame and identify suitable read options or preprocessing steps.

Given dataset: students.csv
ID,Name,Date,Marks
101,"Asha, R",2026-09-01,80
102,Ravi,01/09/2026,
103,Neha,2026-09-03,NA

In [ ]:
#Given code:
import pandas as pd

df = pd.read_csv("students.csv")

print(df)
print(df.dtypes)

### 1. Predict how the quoted comma in "Asha, R" will be handled when the CSV file is read.
  
  The comma inside "Asha, R" will be treated as part of the name, not as a separator between columns. Pandas recognizes the quotation marks and reads the name correctly as Asha, R

### 2. Identify the potential issues with the Date and Marks columns.

  Date column: The dates are written in different formats, such as 2026-09-01 and 01/09/2026. This may cause inconsistent date interpretation if they are not converted into a common date-time format.

  Marks column: It contains missing values, including an empty field. These need to be handled before performing calculations or analysis.

### 3. Rewrite the code to handle missing marks and convert the date column into a suitable date-time format

In [8]:
#code
import pandas as pd

df = pd.read_csv("students.csv")
df["Date"] = pd.to_datetime(df["Date"], format="mixed", dayfirst=True)
df["Marks"] = pd.to_numeric(df["Marks"], errors="coerce")
df["Marks"] = df["Marks"].fillna(df["Marks"].mean())

print(df)
print(df.dtypes)

    ID     Name       Date  Marks
0  101  Asha, R 2026-01-09   80.0
1  102     Ravi 2026-09-01   80.0
2  103     Neha 2026-03-09   80.0
ID                int64
Name                str
Date     datetime64[us]
Marks           float64
dtype: object


### 4. Explain why the value "NA" should be investigated before deciding whether it represents a missing value.
  
  The value "NA" may represent a missing value, but it could also have a different meaning in the dataset. We should check the dataset's context before treating it as missing. If it represents missing data, it can be handled using Pandas missing-value methods. If it has another meaning, converting it to a missing value could lead to incorrect results.

### Q3. A salary column contains the values 25000, 30000, 35000, and 10000000. Explain how the extreme value affects the mean compared with the median. Identify which statistic may better represent a typical employee's salary and justify your choice.

In [ ]:
#Given code:
import pandas as pd

df = pd.DataFrame({
    "Employee": ["A", "B", "C", "D"],
    "Salary": [25000, 30000, 35000, 10000000]
})

print("Mean:", df["Salary"].mean())
print("Median:", df["Salary"].median())

### 1. Predict the mean and median salary.
 
 Mean: Add all salaries and divide by the number of employees.

(25000+30000+35000+10000000)/4
=2522500

Median: Arrange the salaries in ascending order.
25,000, 30,000, 35,000, 10,000,000
take the average of the two middle values.
(30000+35000)/2
=32500
  

### 2. Identify the effect of the extreme salary on the mean.
 
 The extremely high salary of ₹1,00,00,000 increases the mean significantly. As a result, the mean becomes much higher than the salaries of most employees.
 

### 3. Which statistic may better represent a typical employee's salary in this dataset? Justify your answer.
 
 The median may better represent a typical employee's salary because it is less affected by extreme values. In this dataset, most employees earn between ₹25,000 and ₹35,000, so the median of ₹32,500 gives a more representative picture than the mean.
 

### 4. Write code to identify the employee record with the highest salary and explain why the record should be investigated before removing it

In [10]:
import pandas as pd

df = pd.DataFrame({
    "Employee": ["A", "B", "C", "D"],
    "Salary": [25000, 30000, 35000, 10000000]
})
highest_salary = df["Salary"].idxmax()
print(df.loc[highest_salary])


Employee           D
Salary      10000000
Name: 3, dtype: object


### Explanation:

The record should be investigated before removing it to check whether the unusually high salary is correct or due to a data entry error. It may represent a genuine salary, so removing it without verification could lead to loss of valid data and affect the analysis.

# Scenario-based Questions

### Q1. A student assessment dataset contains missing marks, duplicate student records, inconsistent department names such as IT, it, and I.T., and marks stored as text. Develop a cleaning sequence to standardize the data, handle missing values, remove duplicates, and convert marks into a suitable numeric type. State one risk of deleting missing values without investigating why they are missing. 

### 1. Load the CSV file and inspect the dataset using head(), info(), and isnull().sum().

In [14]:
import pandas as pd

df = pd.read_csv("student_assessment_dirty.csv")

display(df.head())
display(df.info())
display(df.isnull().sum())

,Student_ID,Student_Name,Department,Assessment,Marks
0,S101,Asha,IT,Python,85
1,S102,Ravi,it,Python,78
2,S103,Neha,I.T.,Python,92
3,S104,Karan,HR,Python,NaN
4,S105,Meera,HR,Python,Absent


<class 'pandas.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   Student_ID    15 non-null     str  
 1   Student_Name  15 non-null     str  
 2   Department    15 non-null     str  
 3   Assessment    15 non-null     str  
 4   Marks         12 non-null     str  
dtypes: str(5)
memory usage: 732.0 bytes


None

Student_ID      0
Student_Name    0
Department      0
Assessment      0
Marks           3
dtype: int64

### Explanation:

head() displays the first five rows of the dataset.

info() shows the column names, data types, and non-null values.

isnull().sum() counts the missing values in each column.

### 2. Standardize department names so that IT, it, and I.T. are represented consistently. Apply the same approach to Finance.

In [15]:
df["Department"] = df["Department"].str.strip().str.lower()

df["Department"] = df["Department"].replace({
    "it": "IT",
    "i.t.": "IT",
    "finance": "Finance"
})

display(df["Department"].value_counts())

Department
IT         7
hr         4
Finance    4
Name: count, dtype: int64

### Explanation:

str.strip() removes extra spaces.

str.lower() converts department names to lowercase so different versions can be matched.

replace() standardizes the department names.

value_counts() displays the number of records in each department.

### 3. Identify and remove exact duplicate student assessment records.

In [21]:
original_count = len(df)
duplicates = df.duplicated().sum()
print("Number of duplicate records:", duplicates)

df = df.drop_duplicates()

print("Shape after removing duplicates:", df.shape)

Number of duplicate records: 0
Shape after removing duplicates: (14, 5)


### Explanation:

duplicated().sum() counts the exact duplicate rows in the DataFrame.

drop_duplicates() removes duplicate rows while keeping the first occurrence.

shape displays the number of rows and columns remaining after removing duplicates.

### 4. Convert the Marks column into a numeric format, handling "88 marks", "Absent", "NA", and blank entries appropriately.

In [17]:
df["Marks"] = df["Marks"].astype(str).str.replace(" marks", "", regex=False)
df["Marks"] = pd.to_numeric(df["Marks"], errors="coerce")

display(df["Marks"])
display(df["Marks"].isnull().sum())

0      85.0
1      78.0
2      92.0
3       NaN
4       NaN
5       0.0
6      67.0
7      75.0
9       NaN
10     88.0
11     91.0
12      NaN
13     64.0
14    105.0
Name: Marks, dtype: float64

np.int64(4)

### Explanation:

astype(str) converts the values to strings so text can be cleaned.

str.replace() removes " marks" from values such as "88 marks".

pd.to_numeric(..., errors="coerce") converts valid values into numbers and changes values that cannot be converted, such as "Absent" and "NA", into NaN.

isnull().sum() counts the missing values after conversion.

### 5. Investigate missing marks and distinguish absent students from students whose marks have not been entered. State one risk of deleting missing values without investigating why they are missing.

In [ ]:
missing_marks = df[df["Marks"].isnull()]
display(missing_marks)
print("Total missing marks:", df["Marks"].isnull().sum())

,Student_ID,Student_Name,Department,Assessment,Marks
3,S104,Karan,hr,Python,NaN
4,S105,Meera,hr,Python,NaN
9,S109,Sneha,IT,Python,NaN
12,S112,Aman,IT,Python,NaN


Total missing marks: 4


### Explanation:

isnull() identifies records where marks are missing.

Review the other columns in the displayed records to distinguish students marked as Absent from students whose marks have not been entered.

Absent students did not attend the assessment, while missing entries may indicate that their marks have not yet been recorded.

Risk of deleting missing values: If you delete all rows with missing marks without investigating, you may remove students whose marks simply have not been entered, resulting in loss of useful data.

### 6. Identify marks outside the valid range of 0–100 and decide how they should be handled.

In [19]:
invalid_marks = df[(df["Marks"] < 0) | (df["Marks"] > 100)]

display(invalid_marks)
print("Number of invalid marks:", len(invalid_marks))

,Student_ID,Student_Name,Department,Assessment,Marks
14,S114,Kabir,Finance,Python,105.0


Number of invalid marks: 1


### Explanation:

df["Marks"] < 0 finds marks below 0.

df["Marks"] > 100 finds marks above 100.

| means OR, so the condition finds records matching either case.

len(invalid_marks) counts the invalid records.

Decision: Marks should normally be between 0 and 100. Review the invalid records and decide whether to correct them using reliable information or mark them as missing. Do not change values based on guesses.

### 7. Display the cleaned DataFrame and report the number of records removed, missing marks remaining, and invalid marks identified.

In [22]:
display(df)

print("Records removed:", original_count - len(df))
print("Missing marks remaining:", df["Marks"].isnull().sum())
print("Invalid marks identified:", len(invalid_marks))

,Student_ID,Student_Name,Department,Assessment,Marks
0,S101,Asha,IT,Python,85.0
1,S102,Ravi,IT,Python,78.0
2,S103,Neha,IT,Python,92.0
3,S104,Karan,hr,Python,NaN
4,S105,Meera,hr,Python,NaN
5,S106,Arjun,Finance,Python,0.0
6,S107,Priya,Finance,Python,67.0
7,S108,Rahul,IT,Python,75.0
9,S109,Sneha,IT,Python,NaN
10,S110,Varun,hr,Python,88.0


Records removed: 0
Missing marks remaining: 5
Invalid marks identified: 1


### Explanation:

display(df) shows the cleaned DataFrame.

original_count - len(df) calculates how many records were removed. This assumes duplicate removal was the only operation that removed rows.

isnull().sum() counts missing marks remaining after cleaning.

len(invalid_marks) reports how many invalid-mark records you identified in Step 6.